src/kernels/backend_webgl.ts

```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
     return this.compileAndRun(program, [x]) as T;
   }
 
+  conv2dByMatMul(x: Tensor4D, filter: Tensor4D, convInfo: Conv2DInfo):
+      Tensor4D {
+    // Reshapes conv2D input to 2D tensors, uses matMul and then reshape the
+    // result from 2D to 4D.
+    if (!ENV.get('WEBGL_LAZILY_UNPACK') ||
+        !ENV.get('WEBGL_PACK_BINARY_OPERATIONS')) {
+      const $x = this.reshape(x, [1, x.shape[0] * x.shape[1] * x.shape[2],
+          convInfo.inChannels]) as Tensor3D;
+      const $filter = this.reshape(filter,
+          [1, convInfo.inChannels, convInfo.outChannels]) as Tensor3D;
+      return this.batchMatMul($x, $filter, false, false).reshape<Rank.R4>(
+          convInfo.outShape);
+    }
+    
+    const xshape = x.shape;
+    const xTexData = this.texData.get(x.dataId);
+    // We avoid expensive packed reshape by padding row count to next even
+    // number. When number of rows in batch, x.shape[2]), in 2x2 packed texture
+    // layout is odd, the result of packed batchMatMul is the same as it is for
+    // next, even, number of rows in the batch - x.shape[2] + 1.
```

Comment:
This comment is a little bit confusing. Are you saying that if we add an empty row to x and apply pointwise convolution, we can recover the correct result by dropping the last row of the output? Could you rephrase this comment using an example with concrete shapes for `x` and `filter`? ASCII art never hurts, but it's fine if you can just explain this a little more clearly :)

Reply:
Didn't use ascii art but [rephrased the comment](https://github.com/tensorflow/tfjs-core/pull/1495/files#diff-9d35adb9ccc3d36d43672a2b9ba683fbR1650).
Note that the code is simplified with the latest commit, making it easier to read.